src/napi.h

```diff
@@ -0,0 +1,663 @@
+    bool IsNull() const;
+    bool IsBoolean() const;
+    bool IsNumber() const;
+    bool IsString() const;
+    bool IsSymbol() const;
+    bool IsArray() const;
+    bool IsArrayBuffer() const;
+    bool IsTypedArray() const;
+    bool IsObject() const;
+    bool IsFunction() const;
+    bool IsBuffer() const;
+
+    Boolean AsBoolean() const;
+    Number AsNumber() const;
+    String AsString() const;
+    Object AsObject() const;
+    Array AsArray() const;
+    TypedArray AsTypedArray() const;
+    Function AsFunction() const;
+    Buffer AsBuffer() const;
+
+    Boolean ToBoolean() const;
+    Number ToNumber() const;
+    String ToString() const;
+    Object ToObject() const;
+
+  protected:
+    napi_env _env;
+    napi_value _value;
+  };
+
+  class Boolean : public Value {
+  public:
+    static Boolean New(Napi::Env env, bool val);
+
+    Boolean();
+    Boolean(napi_env env, napi_value value);
+
+    operator bool() const;
+
+    bool Value() const;
+  };
+
+  class Number : public Value {
+  public:
+    static Number New(Napi::Env env, double val);
+
+    Number();
+    Number(napi_env env, napi_value value);
+
+    operator int32_t() const;
+    operator uint32_t() const;
+    operator int64_t() const;
+    operator float() const;
+    operator double() const;
+
+    int32_t Int32Value() const;
+    uint32_t Uint32Value() const;
+    int64_t Int64Value() const;
+    float FloatValue() const;
+    double DoubleValue() const;
+  };
+
+  class String : public Value {
+  public:
+    static String New(Napi::Env env, const std::string& value);
+    static String New(Napi::Env env, const std::u16string& value);
+    static String New(Napi::Env env, const char* val, int length = -1);
+    static String New(Napi::Env env, const char16_t* val, int length = -1);
+
+    String();
+    String(napi_env env, napi_value value);
+
+    operator std::string() const;
+    operator std::u16string() const;
+    std::string Utf8Value() const;
+    std::u16string Utf16Value() const;
+  };
+
+  class Object : public Value {
+  public:
+    static Object New(Napi::Env env);
+
+    Object();
+    Object(napi_env env, napi_value value);
+
+    Value operator [](const char* name) const;
+
+    bool Has(const PropertyName& name) const;
+    bool Has(const char* utf8name) const;
+    Value Get(const PropertyName& name) const;
+    Value Get(const char* utf8name) const;
+    void Set(const PropertyName& name, const Value& value);
+    void Set(const char* utf8name, const Value& value);
+    void Set(const char* utf8name, const char* utf8value);
+    void Set(const char* utf8name, bool boolValue);
+    void Set(const char* utf8name, double numberValue);
+
+    void DefineProperty(const PropertyDescriptor& property);
+    void DefineProperties(const std::vector<PropertyDescriptor>& properties);
+    bool InstanceOf(const Function& constructor) const;
+    template<typename T> T* Unwrap() const;
+  };
+
+  class Array : public Value {
```

Comment:
I specifically chose not to do that. The only thing it would gain by inheriting from object would be all the property get/set/etc methods, which would rarely ever be used on an array. In the rare case someone does need to treat an array as an object, it is trivial to call `.AsObject()` on the array first.